### PR TITLE
Remove deprecated `--numbered` flag from four commands

### LINKED ISSUE #6986: Why does `reduce --numbered` cause both the accumulator and the element to become {item, index} records?
### Question

```
〉[a b c d e f] | reduce --fold '' --numbered {|e a| $e.item + $a.item }
fedcba
```
This is what I mean - you have to write `$a.item` as well as `$e.item`, *even though both $a.index and $e.index are always the same!* It just seems very wasteful to me. It also doesn't make sense for the accumulator to have an "index", given that it's persistent across loops. Having only $e be an `{item, index}` record would be better.

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by webbedspace**
# Description

Remove `--numbered` from ~~`for`~~, `each`, `par-each`, `reduce` and `each while`. These all provide indexes (numbering) via the optional second param to their closures.

EDIT: Closes #6986.

# User-Facing Changes

Every command that had `--numbered` listed as "deprecated" in their help docs is affected.

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

### CODE REVIEW

**Comment by fdncred** [comment]
I like this change. It makes it more rusty and intuitive to me.

**Comment by webbedspace** [comment]
Reverted the `for` change because I just noticed that one wasn't deprecated. 🦎

File: crates/nu-command/tests/commands/reduce.rs
Comment (fdncred) [inline crates/nu-command/tests/commands/reduce.rs:62]: Can we keep this test but in the new format? I can't really figure out how to make it work right now.
This prints `longest` as it should but I'm not sure how to get the index.
```
echo one longest three bar | reduce {|it, acc, idx| if ($it | str length) > ($acc | str length) { $it } else { $acc }}
```

**Comment by fdncred** [comment]
Thank you for adding the test. I'm so terrible with reduce. 🤣 Just documenting this here; A way to do it with records so we can `get index` like originally.
```nu
echo one longest three bar | reduce -f {item:'' index:-1} {|it, acc, idx| if ($it | str length) > ($acc.item | str length) { {item:($it) index:($idx)} } else { $acc }} | get index
```

**Review by fdncred — APPROVED** [review]
I think this one is good to go if we're ready for this breaking change.

File: crates/nu-command/src/filters/each.rs
Comment (jntrnr) [inline crates/nu-command/src/filters/each.rs:104]: Should we also change the flag? If the flag is `--keep-empty` now, I wonder if we should make it `--keep-nulls`

Comment (webbedspace) [reply crates/nu-command/src/filters/each.rs:104]: That flag uses `is_nothing()`, which considers `PipelineData::Empty` equivalent to `Value::Nothing`. So, `--keep-empty` is technically just as accurate. (Though, while I think it's more useful for end users to think of `PipelineData::Empty` as being the same as `null`, there are some slight differences here and there that really need to be documented in more detail.)

**Comment by stormasm** [comment]
In our core team meeting on Wednesday we discussed this PR in depth and decided to go with an alternative solution which @jntrnr kindly implemented...

https://github.com/nushell/nushell/pull/7871

**Comment by fdncred** [comment]
I wonder if it's still applicable, now that we have the `enumerate` command, to go back through our code and remove all the `--numbered` parameters on `each`, `for`, and the other commands that implement them? I haven't even tried to use `enumerate` on `for` or any other commands other than `each`.

**Comment by webbedspace** [comment]
Do you want to reopen this PR, then?

**Comment by fdncred** [comment]
It doesn't seem that I can. Github says, "the numbered branch has been deleted".

**Comment by webbedspace** [comment]
OK, try now.

**Comment by fdncred** [comment]
Thanks @webbedspace for working on this. However, I don't think this is what my intent was. This is an example of what I meant using `each` as an example.

1. remove the `--numbered` parameter and it's code. ✅ I believe this part is done. You can'd do `| each --numbered { blah }` anymore. It's ok to keep it in `for` because `for` doesn't accept a pipeline.
2. remove the inferred index closure parameter like `[1 2 3] | each {|item index| print $"item: ($item) index: ($index)"}`. ❌ I don't think this was done.
3. add an example to demonstrate how to use these functions with enumerate now that all then numbering stuff has been removed like `[1 2 3] | enumerate | each {|e| print $"item: ($e.item) index: ($e.index)"}`. ❌ I don't think this was done but I also didn't mention it before.

**Comment by webbedspace** [comment]
>remove the inferred index closure parameter

Ah, right. I personally greatly prefer the inferred index (that I believe I added) because it allows one-liners to be written more quickly like `open x.txt | split row '\n' | each {|e i| $'($i): ($e)' }`, which would be more cumbersomely written as `open x.txt | split row '\n' | enumerate | each { $'($in.index): ($in.item)' }`. Even if a theoretical destructuring syntax was added, the current state of `enumerate` would still make it something like `open x.txt | split row '\n' | enumerate | each {|{index:i item:e}| $'($i): ($e)' }` 

(Though if `enumerate` was changed to produce a 2-list "tuple" instead of an `{index, item}` record, I imagine it'd be more like `open x.txt | split row '\n' | enumerate | each {|[i e]| $'($i): ($e)' }` which is a bit better.)

That being said, if this is consensus, then I guess I'll add all that to this PR shortly.

**Comment by fdncred** [comment]
> I personally greatly prefer the inferred index ...

I tend to lean toward liking the inferred `|item index|` syntax better myself, but when we discussed it further in our meeting on Wednesday, we decided to lean into our composable roots which is where the `enumerate` command came from.

> On the subject of for,  ...

I'm ok with renaming to `--enumerate` in the `for` command.

For the changes to this PR, if you can please make the changes suggested that the core-team agreed to, it would be appreciated.

In the mean time, we can try to address your issue regarding enumerate returning a tuple in #7900

**Comment by webbedspace** [comment]
OK, I've done what was requested. If you can figure out for yourself how to fix the examples whose results I had to comment out, that'd be great, because the error messages I got were *extremely* orthogonal to observed reality.🦜

File: crates/nu-command/src/filters/par_each.rs
Comment (fdncred) [inline crates/nu-command/src/filters/par_each.rs:62]: It would be nice to have this test working but I couldn't figure it out. I tried several different iterations without success.

Comment (webbedspace) [reply crates/nu-command/src/filters/par_each.rs:62]: A shame, as I feel like that one offers a good strategy for using par-each in order-sensitive situations.

Comment (fdncred) [reply crates/nu-command/src/filters/par_each.rs:62]: Let me play some more. At the end of the day, the code snippet works in nushell. It's just that the test doesn't quite work right. So, we could just have a `result: None` with the code snippet in there.

This is what it returns.
```
---- filters::par_each::test::test_examples stdout ----
thread 'filters::par_each::test::test_examples' panicked at 'assertion failed: `(left == right)`
  left: `Error { error: CantFindColumn("item", Span { start: 128, end: 132 }, Span { start: 124, end: 127 }) }`,
 right: `List { vals: [Int { val: 2, span: Span { start: 0, end: 0 } }, Int { val: 4, span: Span { start: 0, end: 0 } }, Int { val: 6, span: Span { start: 0, end: 0 } }], span: Span { start: 0, end: 0 } }`: The example result differs from the expected value', crates\nu-command\src\example_test.rs:231:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    filters::par_each::test::test_examples

test result: FAILED. 398 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.27s

error: test failed, to rerun pass `-p nu-command --lib`
```

Comment (fdncred) [reply crates/nu-command/src/filters/par_each.rs]: I think I figured it out. Some commands weren't in `example_test.rs`.

**Comment by rgwood** [comment]
I'm having some trouble with this one. It's broken some scripts in `nu_scripts` and I'm not sure how to fix them. This probably needs a breaking change documentation in the blog post: https://github.com/nushell/nushell.github.io/pull/770/files

How can I get the index when using `each` after this change? This breaks:

```bash
〉ls | each { |f n| print $n }
Error: nu::shell::eval_block_with_input (link)

  × Eval block failed with pipeline input
   ╭─[entry #2:1:1]
 1 │ ls | each { |f n| print $n }
   · ─┬
   ·  ╰── source value
   ╰────

Error: nu::shell::variable_not_found (link)

  × Variable not found
   ╭─[entry #2:1:1]
 1 │ ls | each { |f n| print $n }
   ·                         ─┬
   ·                          ╰── variable not found
   ╰────
```

**Comment by fdncred** [comment]
@rgwood `help each` reveals this which may help.
```
Iterate over each element, producing a list showing indexes of any 2s
  > [1 2 3] | enumerate | each {|e| if $e.item == 2 { $"found 2 at ($e.index)!"} }
```

**Pull request merged**
